File: 2216-DeleteTheMiddleNodeOfALinkedList/2216-DeleteTheMiddleNodeOfALinkedList.py
# Last updated: 9/14/2025, 9:08:53 AM
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or  not head.next:
            return None
        fast,slow,prev=head,head,head

        while fast and fast.next:
            prev=slow
            slow=slow.next
            fast=fast.next.next
        prev.next=slow.next

        return head

File: 2216-DeleteTheMiddleNodeOfALinkedList/2216-DeleteTheMiddleNodeOfALinkedList.java
// Last updated: 9/14/2025, 9:08:47 AM
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode deleteMiddle(ListNode head) {
        ListNode tortoise = head;
        ListNode hare = head;
        ListNode prev = null;
        while(hare != null && hare.next!=null){
            prev = tortoise;
            tortoise = tortoise.next;
            hare = hare.next.next;
        }
        if(prev == null || prev.next == null) return null;
        prev.next = tortoise.next;
        tortoise = null;
        return head;
    }
}